File: pascal.js
/**
 * Funcion para calcular el triangulo de pascal.
 * Pasado un cierto level, calculara siempre y cuando level sea un numero entre 0 a 25
 * 
 * @param {number} level 
 * @returns {void}
 */
export default function pascalTriangle(level = 2) {
    if (typeof level !== "number") {
        console.log("El parametro level tiene que ser un numero")
        return
    }

    if (level <= 0) {
        console.log("El nivel ingresado es imposible de calcular")
        return
    }

    if (level > 25) {
        console.log("El nivel excede el limite de 25, la ejecucion se cancelo")
        return
    }

    const firstLevel = [1]
    const secondLevel = [1, 1]
    let prevLevel = secondLevel

    printLevel(firstLevel)
    if (level <= 1) return "1"

    printLevel(secondLevel)
    if (level <= 2) return "1 1"

    for (let i = 0; i < level - 2; i++) {
        const currLevel = calculateLevel(prevLevel)
        printLevel(currLevel)
        prevLevel = currLevel
    }
}

// imprime el nivel con numeros espaciados
function printLevel(level) {
    console.log(level.join(" ").toString())
}

// calcula en nivel segun el nivel previo
function calculateLevel(prevLevel) {
    const level = []

    for (let i = 0; i <= prevLevel.length; i++) {
        const curr = prevLevel[i] || undefined
        const prev = prevLevel[i - 1] || undefined

        // en la primera iteracion no hay anterior, asi que solo le ponemos el valor actual = 1
        if (!prev) {
            level.push(curr)

            // en la ultima iteracion no hay un valor actual, asi que solo le ponemos el valor anterior = 1
        } else if (!curr) {
            level.push(prev)

            // calculamos el valor con el numero actual y el previo.
        } else {
            const sum = curr + prev
            level.push(sum)
        }

    }

    level.push[1]
    return level
}

File: Repaso/Semana 1/pascal.py
"""
Función para calcular el triángulo de Pascal.
Dado un cierto nivel, calculará siempre y cuando
level sea un número entre 0 y 25.
"""
def pascal_triangle(level=2):
    # Validación de tipo
    if not isinstance(level, int):
        print("El parametro level tiene que ser un numero")
        return

    # Validación de rango
    if level <= 0:
        print("El nivel ingresado es imposible de calcular")
        return

    if level > 25:
        print("El nivel excede el limite de 25, la ejecucion se cancelo")
        return

    first_level = [1]
    second_level = [1, 1]

    prev_level = second_level

    print_level(first_level)

    if level <= 1:
        return

    print_level(second_level)

    if level <= 2:
        return

    # Generar los siguientes niveles
    for _ in range(level - 2):
        curr_level = calculate_level(prev_level)
        print_level(curr_level)
        prev_level = curr_level


# Imprime el nivel con números espaciados
def print_level(level):
    print(" ".join(map(str, level)))

# Calcula el siguiente nivel según el nivel previo
def calculate_level(prev_level):

    level = []

    for i in range(len(prev_level) + 1):

        curr = prev_level[i] if i < len(prev_level) else None
        prev = prev_level[i - 1] if i - 1 >= 0 else None

        # Primera iteración
        if prev is None:
            level.append(curr)

        # Última iteración
        elif curr is None:
            level.append(prev)

        # Suma de elementos intermedios
        else:
            level.append(curr + prev)

    return level

# main
size = int(input("Ingrese el tamaño que desea del triangulo: "))
pascal_triangle(size)